// angular-toastify/esm2015/lib/toastify-toast/toastify-toast.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/toastify-toast/toastify-toast.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, EventEmitter, HostListener, Input, NgZone, Output } from '@angular/core';
import { ToastType } from '../toast-type';
import { Toast } from '../toast';
export class ToastifyToastComponent {
    /**
     * @param {?} _cd
     * @param {?} _zone
     */
    constructor(_cd, _zone) {
        this._cd = _cd;
        this._zone = _zone;
        this.autoClose = 5000;
        this.hideProgressBar = false;
        this.pauseOnHover = true;
        this.pauseOnVisibilityChange = true;
        this.closeOnClick = true;
        this.dismissEvent = new EventEmitter();
        this.ToastType = ToastType;
        this.running = false;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.autoCloseRemaining = this.autoClose;
        this.startTime = this.toast.time;
        // Do not start timer when toast is prompted while window is out of focus
        if (this.handleVisibilityChange && document.visibilityState === 'visible') {
            this.startCloseTimer();
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.clearTimerTimeout();
    }
    /**
     * @return {?}
     */
    startCloseTimer() {
        if (this.running || !this.autoClose) {
            return;
        }
        this.startTime = new Date().getTime();
        this.running = true;
        this.autoDismissTimeout = this._zone.runOutsideAngular((/**
         * @return {?}
         */
        () => setTimeout((/**
         * @return {?}
         */
        () => {
            this._zone.run((/**
             * @return {?}
             */
            () => {
                this.dismissEvent.emit();
                this._cd.markForCheck();
            }));
        }), this.autoCloseRemaining)));
    }
    /**
     * @return {?}
     */
    pauseCloseTimer() {
        this.running = false;
        this.clearTimerTimeout();
        // Calculate the elapsed time, subtract remaining time
        this.pauseTime = new Date().getTime();
        /** @type {?} */
        const elapsed = this.pauseTime - this.startTime;
        this.autoCloseRemaining -= elapsed;
    }
    /**
     * @return {?}
     */
    clearTimerTimeout() {
        if (this.autoDismissTimeout !== undefined) {
            clearTimeout(this.autoDismissTimeout);
        }
    }
    /**
     * @return {?}
     */
    handleDismissButtonAction() {
        if (this.closeOnClick) {
            return; // Let the other event handle the dismissal
        }
        this.clearTimerTimeout();
        this.dismissEvent.emit();
    }
    /**
     * @return {?}
     */
    handleHostClick() {
        if (this.closeOnClick) {
            this.clearTimerTimeout();
            this.dismissEvent.emit();
        }
    }
    /**
     * @return {?}
     */
    handleMouseEnter() {
        if (this.pauseOnHover) {
            this.pauseCloseTimer();
        }
    }
    /**
     * @return {?}
     */
    handleMouseLeave() {
        if (this.pauseOnHover) {
            this.startCloseTimer();
        }
    }
    /**
     * @return {?}
     */
    handleVisibilityChange() {
        if (!this.pauseOnVisibilityChange) {
            return;
        }
        if (document.visibilityState !== 'visible') {
            this.pauseCloseTimer();
            this._cd.detectChanges();
        }
        else {
            this.startCloseTimer();
        }
    }
}
ToastifyToastComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'lib-toastify-toast',
                template: "<div role=\"alert\" class=\"toast-body\">\n    <div class=\"icon-container\">\n        <span *ngIf=\"iconLibrary == 'material'\" [ngSwitch]=\"toast.type\">\n            <i class=\"material-icons\" *ngSwitchCase=\"ToastType.info\">info_outline</i>\n            <i class=\"material-icons\" *ngSwitchCase=\"ToastType.default\">info_outline</i>\n            <i class=\"material-icons\" *ngSwitchCase=\"ToastType.warning\">warning_outline</i>\n            <i class=\"material-icons\" *ngSwitchCase=\"ToastType.error\">error_outline</i>\n            <i class=\"material-icons\" *ngSwitchCase=\"ToastType.success\">done</i>\n        </span>\n        <span *ngIf=\"iconLibrary == 'font-awesome'\" [ngSwitch]=\"toast.type\">\n            <i *ngSwitchCase=\"ToastType.info\" class=\"fa fa-info\"></i>\n            <i *ngSwitchCase=\"ToastType.default\" class=\"fa fa-info\"></i>\n            <i *ngSwitchCase=\"ToastType.warning\" class=\"fa fa-exclamation-triangle\"></i>\n            <i *ngSwitchCase=\"ToastType.error\" class=\"fa fa-exclamation\"></i>\n            <i *ngSwitchCase=\"ToastType.success\" class=\"fa fa-check\"></i>\n        </span>\n    </div>\n    <div class=\"toast-container\">\n        <span>{{toast.message}}</span>\n    </div>\n</div>\n<button (click)=\"handleDismissButtonAction()\" class=\"close-button close-button--{{ToastType[toast.type]}}\" type=\"button\" aria-label=\"close\">\u2716</button>\n<div *ngIf=\"!hideProgressBar\" class=\"progress-bar progress-bar--{{ToastType[toast.type]}}\"\n    [style.animation-play-state]=\"running? 'running' : 'paused'\" [style.animation-duration]=\"this.autoClose + 'ms'\" style=\"opacity: 1;\"></div>",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [".toast{border-radius:1px;box-shadow:0 1px 10px 0 rgba(0,0,0,.1),0 2px 15px 0 rgba(0,0,0,.05);box-sizing:border-box;cursor:pointer;direction:ltr;display:flex;font-family:sans-serif;justify-content:space-between;margin-bottom:1rem;max-height:800px;min-height:64px;overflow:hidden;padding:8px;position:relative}.toast--rtl{direction:rtl}.toast--default{background:#fff;color:#aaa}.toast--info{background:#3498db}.toast--success{background:#07bc0c}.toast--warning{background:#f1c40f}.toast--error{background:#e74c3c}.toast-body{flex:1;margin:auto 0}@media only screen and (max-width:480px){.toast{margin-bottom:0}}.close-button{align-self:flex-start;background:transparent;border:none;color:#fff;cursor:pointer;font-size:14px;font-weight:700;opacity:.7;outline:none;padding:0;transition:.3s ease}.close-button--default{color:#000;opacity:.3}.close-button:focus,.close-button:hover{opacity:1}@-webkit-keyframes trackProgress{0%{transform:scaleX(1)}to{transform:scaleX(0)}}@keyframes trackProgress{0%{transform:scaleX(1)}to{transform:scaleX(0)}}.progress-bar{background-color:hsla(0,0%,100%,.7);bottom:0;height:5px;left:0;opacity:.7;position:absolute;transform-origin:left;width:100%;z-index:9999}.progress-bar,.progress-bar--animated{-webkit-animation:trackProgress linear 1 forwards;animation:trackProgress linear 1 forwards}.progress-bar--controlled{transition:transform .2s}.progress-bar--rtl{left:auto;right:0;transform-origin:right}.progress-bar--default{background:linear-gradient(90deg,#4cd964,#5ac8fa,#007aff,#34aadc,#5856d6,#ff2d55)}.icon-container,.toast-body{vertical-align:middle}.icon-container{display:inline-block;width:25px}.icon-container span{display:inline-block;margin-right:.5rem;text-align:center;width:20px}.icon-container span i{vertical-align:middle}.fa,.material-icons{font-size:18px}.toast-container{display:inline-block;vertical-align:middle;width:calc(100% - 25px)}"]
            }] }
];
/** @nocollapse */
ToastifyToastComponent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: NgZone }
];
ToastifyToastComponent.propDecorators = {
    autoClose: [{ type: Input }],
    hideProgressBar: [{ type: Input }],
    pauseOnHover: [{ type: Input }],
    pauseOnVisibilityChange: [{ type: Input }],
    closeOnClick: [{ type: Input }],
    toast: [{ type: Input }],
    iconLibrary: [{ type: Input }],
    dismissEvent: [{ type: Output }],
    handleHostClick: [{ type: HostListener, args: ['click',] }],
    handleMouseEnter: [{ type: HostListener, args: ['mouseenter',] }],
    handleMouseLeave: [{ type: HostListener, args: ['mouseleave',] }],
    handleVisibilityChange: [{ type: HostListener, args: ['document:visibilitychange',] }]
};
if (false) {
    /** @type {?} */
    ToastifyToastComponent.prototype.autoClose;
    /** @type {?} */
    ToastifyToastComponent.prototype.hideProgressBar;
    /** @type {?} */
    ToastifyToastComponent.prototype.pauseOnHover;
    /** @type {?} */
    ToastifyToastComponent.prototype.pauseOnVisibilityChange;
    /** @type {?} */
    ToastifyToastComponent.prototype.closeOnClick;
    /** @type {?} */
    ToastifyToastComponent.prototype.toast;
    /** @type {?} */
    ToastifyToastComponent.prototype.iconLibrary;
    /** @type {?} */
    ToastifyToastComponent.prototype.dismissEvent;
    /**
     * @type {?}
     * @private
     */
    ToastifyToastComponent.prototype.autoDismissTimeout;
    /**
     * @type {?}
     * @private
     */
    ToastifyToastComponent.prototype.autoCloseRemaining;
    /**
     * @type {?}
     * @private
     */
    ToastifyToastComponent.prototype.pauseTime;
    /**
     * @type {?}
     * @private
     */
    ToastifyToastComponent.prototype.startTime;
    /** @type {?} */
    ToastifyToastComponent.prototype.ToastType;
    /** @type {?} */
    ToastifyToastComponent.prototype.running;
    /**
     * @type {?}
     * @private
     */
    ToastifyToastComponent.prototype._cd;
    /**
     * @type {?}
     * @private
     */
    ToastifyToastComponent.prototype._zone;
}
