// angular-toastify/esm2015/lib/toast.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/toast.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, EventEmitter } from '@angular/core';
import { Toast } from './toast';
import { ToastType } from './toast-type';
import * as i0 from "@angular/core";
export class ToastService {
    constructor() {
        this.toastAddedEvent = new EventEmitter();
        this.dismissAllEvent = new EventEmitter();
    }
    /**
     * @return {?}
     */
    dismissAll() {
        this.dismissAllEvent.emit();
    }
    /**
     * @param {?} message
     * @return {?}
     */
    info(message) {
        /** @type {?} */
        const toast = new Toast(message, ToastType.info);
        this.toastAddedEvent.emit(toast);
    }
    /**
     * @param {?} message
     * @return {?}
     */
    success(message) {
        /** @type {?} */
        const toast = new Toast(message, ToastType.success);
        this.toastAddedEvent.emit(toast);
    }
    /**
     * @param {?} message
     * @return {?}
     */
    warn(message) {
        /** @type {?} */
        const toast = new Toast(message, ToastType.warning);
        this.toastAddedEvent.emit(toast);
    }
    /**
     * @param {?} message
     * @return {?}
     */
    error(message) {
        /** @type {?} */
        const toast = new Toast(message, ToastType.error);
        this.toastAddedEvent.emit(toast);
    }
    /**
     * @param {?} message
     * @return {?}
     */
    default(message) {
        /** @type {?} */
        const toast = new Toast(message, ToastType.default);
        this.toastAddedEvent.emit(toast);
    }
}
ToastService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
ToastService.ctorParameters = () => [];
/** @nocollapse */ ToastService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ToastService_Factory() { return new ToastService(); }, token: ToastService, providedIn: "root" });
if (false) {
    /** @type {?} */
    ToastService.prototype.toastAddedEvent;
    /** @type {?} */
    ToastService.prototype.dismissAllEvent;
}
